dataplane/ebpf/src/egress/icmp.rs

```diff
@@ -0,0 +1,93 @@
+use core::mem;
+
+use aya_bpf::{
+    bindings::TC_ACT_PIPE,
+    helpers::bpf_csum_diff,
+    programs::TcContext,
+};
+use aya_log_ebpf::info;
+
+use crate::{
+    bindings::{iphdr, icmphdr},
+    utils::{csum_fold_helper, ip_from_int, ptr_at, ETH_HDR_LEN, IP_HDR_LEN},
+    CONNTRACK,
+};
+
+const ICMP_HDR_LEN: usize = mem::size_of::<icmphdr>();
+const ICMP_PROTO_TYPE_UNREACH: u8 = 3;
+
+pub fn handle_icmp_egress(ctx: TcContext) -> Result<i32, i64> {
+    let ip_hdr: *mut iphdr = unsafe { ptr_at(&ctx, ETH_HDR_LEN) }?;
+
+    let icmp_header_offset = ETH_HDR_LEN + IP_HDR_LEN;
+
+    let icmp_hdr: *mut icmphdr = unsafe { 
+        ptr_at(
+            &ctx,
+            icmp_header_offset
+        )?
+    };
+
+    // We only care about redirecting port unreachable messages currently so a 
+    // UDP client can tell when the server is shutdown
+    if unsafe { (*icmp_hdr).type_ } != ICMP_PROTO_TYPE_UNREACH { 
+        return Ok(TC_ACT_PIPE);
+    }   
+
+    let dest_addr = unsafe { (*ip_hdr).daddr };
+    
+    let new_src = unsafe { CONNTRACK.get(&dest_addr) }.ok_or(TC_ACT_PIPE)?;
+
+    let daddr_dot_dec = ip_from_int(unsafe { (*ip_hdr).daddr });
+    info!(
+        &ctx,
+        "Received a ICMP Unreachable packet destined for svc ip: {}.{}.{}.{}",
+        daddr_dot_dec[0],
+        daddr_dot_dec[1],
+        daddr_dot_dec[2],
+        daddr_dot_dec[3],
+    );
```

Comment:
It occurs to me that I added this for convenience (and boy is it an improvement) but we can probably add one more wrapper on top of this to clean it up further so we don't have to enumerate the indices in every call. Doesn't need to be done in this PR though.

Reply:
Totally agreed let me make an issue for it.

Reply:
https://github.com/Kong/blixt/issues/71